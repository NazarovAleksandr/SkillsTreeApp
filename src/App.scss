@import './variables.scss';

#root, body, html {
    height: 100%;
    margin: 0;
    padding: 0;
    overflow: hidden;
    background-color: black;
}

.App {
    font-family: Arial, Helvetica, sans-serif;
    height: 100%;
    display: flex;
    margin: auto;
    max-width: 1650px;

    .layout {
        display: flex;
        flex-direction: column;
        flex-grow: 1;
        margin: 3%;

        .pages {
            flex-grow: 1;
            display: flex;
            overflow: auto;
        }
    }

    .menu {
        border-bottom: 2px solid $mainBorderColor;
        margin-bottom: 10px;
        text-align: center;

        > a {
            display: inline-block;
            margin-right: 20px;
            padding: 5px;
            color: #faa318;
            text-decoration: none;
            width: 100px;

            &:last-child {
                margin-right: 0;
            }

            &.active, &:hover {
                box-shadow: 0px 2px #faa318;
            }
        }
    }
}