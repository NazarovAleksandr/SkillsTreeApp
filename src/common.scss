@import './variables.scss';

.rune-socket {
    width: 64px;
    height: 64px;
    display: inline-block;
    border: 2px solid;
    border-radius: 40px;
    background: #171717;
    vertical-align: middle;
    box-shadow: 0 2px #000;
    margin: 8px;

    &.unique {
        color: $uniqueRarity;
    }

    &.common {
        color: $commonRarity;
    }

    &.rare {
        color: $rareRarity;
    }
}

.rune {
    &-image {
        width: 64px;
        height: 64px;
    }

    &.used {
        background-color: yellow;
    }
}

.block-title {
    border-bottom: 2px solid $mainBorderColor;
    margin-bottom: 10px;
    padding: 10px;
    color: #faa318;

    .title {
        padding: 6px;
        display: inline-block;
        text-transform: uppercase;
    }
}

.action {
    color: rgb(250, 163, 24);
    border: 1px solid #545454;
    display: inline-block;
    padding: 5px;
    box-shadow: 2px 2px #000;
    text-align: center;
    cursor: default;
    position: relative;
    user-select: none;

    &.disabled {
        color: gray;
    }

    &:hover:not(.disabled) {
        border-color: rgb(250, 163, 24);
        cursor: pointer;
    }
}

@keyframes drop-splash {
	0% {
        width: 70px;
        height: 70px;
		opacity: 1;
		transform: scale3d(0.5, 0.5, 1);
	}
	100% {
		opacity: 0;
        transform: scale3d(1.1, 1.1, 1);
        width: 70px;
        height: 70px;
	}
}

@keyframes drop-splash-2 {
	0% {
        width: 70px;
        height: 70px;
		opacity: 1;
		transform: scale3d(0.5, 0.5, 1);
	}
	50%, 100% {
        width: 70px;
        height: 70px;
		opacity: 0;
		transform: scale3d(1.2, 1.2, 1);
	}
}

@keyframes rotation {
    100% {
        transform: rotate(360deg);
    }
}