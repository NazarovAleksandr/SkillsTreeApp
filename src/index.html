<!DOCTYPE html>
<html>
  <head>
    <meta charset="utf-8"/>
    <meta name="viewport" content="initial-scale=0.6, shrink-to-fit=yes">
    <title>Game proof-of-concept</title>
  </head>
  <body>
    <div id="root"></div>
</html>