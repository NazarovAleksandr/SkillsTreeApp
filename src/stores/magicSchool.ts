import { observable, action, IObservableArray } from 'mobx';
import * as Models from '../components/magicSchools/models';
import * as utils from '../utils';

export class MagicSchoolsStore {
    private schools: IObservableArray<Models.IMagicSchool> = observable([]);

	@action.bound
    addSchool(school: Models.IMagicSchool) {
        this.schools.push(school);
    }

	generateSchool(name = 'New School') {
	    return {
	        id: utils.uids.getUID(),
	        name,
	    };
	}

	@action.bound
	removeSchool(schoolId: string) {
	    const idx = this.schools.findIndex((school) => school.id === schoolId);
	    if (idx > -1) {
	        this.schools.splice(idx, 1);
	    }
	}

	getSchools() {
	    return this.schools;
	}

	getSchool(id: string) {
	    return this.schools.find((school) => school.id === id);
	}
}

export class MagicSchoolsUIStateStore {
	@observable private selectedSchoolId?: string;

	@observable private inEditionState: boolean;

	@action.bound
	setActiveSchool(schoolId: string) {
	    this.endSchoolEdition();
	    this.selectedSchoolId = schoolId;
	}

	@action.bound
	startSchoolEdition() {
	    if (this.selectedSchoolId) {
	        this.inEditionState = true;
	    }
	}

	@action.bound
	endSchoolEdition() {
	    this.inEditionState = false;
	}

	@action.bound
	deselectSchool() {
	    this.selectedSchoolId = undefined;
	    this.endSchoolEdition();
	}

	getSelectedSchoolId() {
	    return this.selectedSchoolId;
	}

	isInEditionState() {
	    return this.inEditionState;
	}
}
