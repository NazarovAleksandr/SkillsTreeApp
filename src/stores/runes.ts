import {
    observable, action, IObservableArray,
} from 'mobx';
import * as Models from '../components/runes/models';
import * as utils from '../utils';

export class RunesStore {
    private runes: IObservableArray<Models.IRune> = observable([]);

    private runeRarities: Models.Rarity[] = [Models.Rarity.common, Models.Rarity.rare, Models.Rarity.unique];

    private runeTypes: Models.RuneTypes[] = [
        Models.RuneTypes.attack,
        Models.RuneTypes.defence,
        Models.RuneTypes.heal,
        Models.RuneTypes.magic,
        Models.RuneTypes.utility,
    ];

    private propertyTypes: Models.RunePropertyTypes[] = [
        Models.RunePropertyTypes.Evade,
        Models.RunePropertyTypes.Hp,
        Models.RunePropertyTypes.Mp,
        Models.RunePropertyTypes.Power,
        Models.RunePropertyTypes.Shield,
    ];

    @action.bound
    public addRune(rune: Models.IRune) {
        this.runes.push(rune);
    }

    public getRunes() {
        return this.runes;
    }

    public getRune(id: string) {
        return this.runes.find((rune) => rune.id === id);
    }

    public generateRune(): Models.IRune {
        const newRarity = this.getRandomEntity(this.runeRarities);
        const newType = this.getRandomEntity(this.runeTypes);

        let maxPropCount;
        let minPropCount;
        switch (newRarity) {
        case Models.Rarity.common:
            minPropCount = 1;
            maxPropCount = 1;
            break;
        case Models.Rarity.rare:
            minPropCount = 1;
            maxPropCount = 2;
            break;
        case Models.Rarity.unique:
            minPropCount = 2;
            maxPropCount = 4;
            break;
        default:
            minPropCount = 1;
            maxPropCount = 1;
            break;
        }

        return {
            id: utils.uids.getUID(),
            type: newType,
            rarity: newRarity,
            properties: this.generateProperties(minPropCount, maxPropCount),
            image: this.generateImage(newType),
        };
    }

    private generateImage = (type: Models.RuneTypes) => {
        let imgType = '';
        switch (type) {
        case Models.RuneTypes.attack:
            imgType = 'precision';
            break;
        case Models.RuneTypes.defence:
            imgType = 'domination';
            break;
        case Models.RuneTypes.heal:
            imgType = 'resolve';
            break;
        case Models.RuneTypes.magic:
            imgType = 'sorcery';
            break;
        case Models.RuneTypes.utility:
            imgType = 'inspiration';
            break;
        default:
            imgType = 'inspiration';
            break;
        }

        const typeImages = utils.images[imgType];
        const picNum = utils.randomizer.getRandomInt(0, typeImages.length - 1);

        return typeImages[picNum];
    }

    private generateProperties(minPropCount: number, maxPropCount: number) {
        const newProperties: Models.IProperty[] = [];
        for (let i = 0; i < minPropCount; i += 1) {
            newProperties.push({
                name: this.getRandomEntity(this.propertyTypes),
                value: utils.randomizer.getRandomInt(-100, 100),
            });
        }

        const randAdditionalPropCount = utils.randomizer.getRandomInt(0, maxPropCount - minPropCount);
        for (let i = 0; i < randAdditionalPropCount; i += 1) {
            newProperties.push({
                name: this.getRandomEntity(this.propertyTypes),
                value: utils.randomizer.getRandomInt(-100, 100),
            });
        }

        return newProperties;
    }

    private getRandomEntity = <T>(list: T[]) => list[utils.randomizer.getRandomInt(0, list.length - 1)];
}
