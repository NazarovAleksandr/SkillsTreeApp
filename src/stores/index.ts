import * as magicSchools from './magicSchool';
import * as runes from './runes';
import * as skillTrees from './skilltrees';
import * as tooltipState from './tooltipState';

export {
    magicSchools,
    runes,
    skillTrees,
    tooltipState
}