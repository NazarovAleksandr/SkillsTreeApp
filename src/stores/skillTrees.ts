import {
    observable, action, computed, IObservableArray,
} from 'mobx';
import * as RuneModels from '../components/runes/models';
import * as utils from '../utils';
import { ISkillsTreeNode } from '../components/skillsTree/models';


export class SkillTreesStore {
    @observable private trees: IObservableArray<ISkillsTreeNode>;

    @observable private schoolIdToTree: {[key: string]: ISkillsTreeNode}

    constructor() {
        this.trees = observable([]);
        this.schoolIdToTree = {};
    }

    @action.bound
    public removeTree(schoolId: string) {
        const treeToRemove = this.schoolIdToTree[schoolId];
        const idx = this.trees.findIndex((tree) => tree === treeToRemove);
        if (idx > -1) {
            this.trees.splice(idx, 1);
            this.schoolIdToTree[schoolId] = undefined;
        }
    }

    @action.bound
    public addTree(schoolId: string, tree: ISkillsTreeNode) {
        this.schoolIdToTree[schoolId] = tree;
        this.trees.push(this.schoolIdToTree[schoolId]);
    }

    @action.bound
    public createTree(schoolId: string, isRoot = false) {
        const newTree: ISkillsTreeNode = { id: utils.uids.getUID() };
        if (isRoot) {
            newTree.isRoot = true;
        }
        this.schoolIdToTree[schoolId] = newTree;
        this.trees.push(this.schoolIdToTree[schoolId]);
    }

    @computed
    get totalNodesCount() {
        let count = 0;
        this.trees.forEach((tree) => {
            count += 1;
            if (tree.children) {
                tree.children.forEach((child) => {
                    count += this.getLength(child);
                });
            }
        });
        return count;
    }

    @computed
    get attachedRunes() {
        let runes: RuneModels.IRune[] = [];
        this.trees.forEach((tree) => {
            runes = [...runes, ...this.getRunes(tree)];
        });
        return new Set(runes);
    }

    public getNode(nodeId: string): ISkillsTreeNode {
        let found;
        this.trees.find((node) => {
            found = this.findNode(node, nodeId);
            return found;
        });
        return found;
    }

    private findNode(node: ISkillsTreeNode, id: string) {
        let found;
        if (node.id === id) {
            found = node;
        }
        if (!found && node.children) {
            node.children.find((child) => {
                found = this.findNode(child, id);
                return found;
            });
        }
        return found;
    }

    private getRunes(tree: ISkillsTreeNode) {
        let runes: RuneModels.IRune[] = [];
        if (tree.attachedRune) {
            runes.push(tree.attachedRune);
        }
        if (tree.children) {
            tree.children.forEach((child) => {
                runes = [...runes, ...this.getRunes(child)];
            });
        }
        return runes;
    }

    private getLength(tree: ISkillsTreeNode) {
        let count = 1;
        if (tree.children) {
            tree.children.forEach((child) => {
                count += this.getLength(child);
            });
        }
        return count;
    }

    public getTree(schoolId: string) {
        return this.schoolIdToTree[schoolId];
    }
}
