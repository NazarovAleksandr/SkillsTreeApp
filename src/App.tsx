import * as React from 'react';
import './App.scss';
import './common.scss';
import { observer } from "mobx-react";
import AboutPage from './pages/about';
import UserSkillsPage from './pages/userSkills';
import { BrowserRouter as Router, Route, NavLink } from "react-router-dom";
import { TooltipContext } from './contexts/tooltip';
import * as stores from './stores';
import { Tooltip } from './components/tooltip/tooltip';

const tooltipStateStore = new stores.tooltipState.TooltipStateStore();
let basename = process.env.BASENAME || '/'

window.addEventListener('scroll', () => {
    tooltipStateStore.hide();
}, {
    capture: true,
    passive: true
});

const App = () => {
    return (
        <Router basename={basename}>
            <TooltipContext.Provider value={tooltipStateStore}>
                <div className="App">
                    <div className="layout">
                        <div className="menu">
                            <NavLink exact to="/" activeClassName='active'>About</NavLink>
                            <NavLink to="/skills" activeClassName='active'>Skills</NavLink>
                        </div>
                        <div className="pages">
                            <Route path="/skills" component={UserSkillsPage} />
                            <Route exact path="/" component={AboutPage} />
                        </div>
                    </div>
                </div>
            </TooltipContext.Provider>
            <Tooltip 
                position={tooltipStateStore.tooltipPosition}
                content={tooltipStateStore.tooltipContent}
                node={tooltipStateStore.hostNode}
                visible={tooltipStateStore.isVisible}
                mouseOutCallback={tooltipStateStore.mouseOutCallback}
            />
        </Router>
    );
}

export default observer(App);

/*

TODO

1) Рефакторинг

*/