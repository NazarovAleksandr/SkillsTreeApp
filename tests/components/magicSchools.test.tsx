import * as React from 'react';
import { shallow, mount, ReactWrapper } from 'enzyme';
import MagicSchools from '../../src/components/magicSchools/magicSchools';
import * as stores from '../../src/stores';

let magicSchoolsStore: stores.magicSchools.MagicSchoolsStore;
let uiStateStore: stores.magicSchools.MagicSchoolsUIStateStore;

beforeEach(() => {
    magicSchoolsStore = new stores.magicSchools.MagicSchoolsStore();
    uiStateStore = new stores.magicSchools.MagicSchoolsUIStateStore();
});

describe('MagicSchools', () => {
    test('Renders', () => {
        let school = {id: '1', name: '1'};
        let school2 = {id: '2', name: '2'};
        magicSchoolsStore.addSchool(school);
        magicSchoolsStore.addSchool(school2);

        const component = shallow(<MagicSchools schoolsStore={magicSchoolsStore} uiStateStore={uiStateStore} ></MagicSchools>);

        expect(component).toMatchSnapshot();
    });
    test('Adds new school', () => {
        const component = shallow<MagicSchools>(<MagicSchools schoolsStore={magicSchoolsStore} uiStateStore={uiStateStore} ></MagicSchools>);

        component.instance().addNewItem();

        expect(magicSchoolsStore.getSchools()).toHaveLength(1);
        expect(uiStateStore.isInEditionState()).toEqual(true);
    });
    test('Removes school', () => {
        const component = shallow<MagicSchools>(<MagicSchools schoolsStore={magicSchoolsStore} uiStateStore={uiStateStore} ></MagicSchools>);
        let school = {id: '1', name: '1'};
        magicSchoolsStore.addSchool(school);

        uiStateStore.setActiveSchool('1');
        component.instance().deleteItem();

        expect(magicSchoolsStore.getSchools()).toHaveLength(0);
        expect(uiStateStore.getSelectedSchoolId()).toBeFalsy();
    });
    test('Makes school editable', () => {
        const component = shallow<MagicSchools>(<MagicSchools schoolsStore={magicSchoolsStore} uiStateStore={uiStateStore} ></MagicSchools>);
        let school = {id: '1', name: '1'};
        magicSchoolsStore.addSchool(school);

        uiStateStore.setActiveSchool('1');
        component.instance().startEdition();

        expect(uiStateStore.isInEditionState()).toEqual(true);
        expect(component).toMatchSnapshot('One school is in edition state');
    });
    test('Starts edition on select', () => {
        const component = shallow<MagicSchools>(<MagicSchools schoolsStore={magicSchoolsStore} uiStateStore={uiStateStore} ></MagicSchools>);
        let school = {id: '1', name: '1'};
        
        component.instance().onEditionEnd(school, 'new');

        expect(school.name).toEqual('new');
        expect(uiStateStore.isInEditionState()).toEqual(false);
    });
});