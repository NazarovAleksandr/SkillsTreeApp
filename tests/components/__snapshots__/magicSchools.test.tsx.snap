// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicSchools Makes school editable: One school is in edition state 1`] = `
<div
  className="schools-block"
>
  <div
    className="block-title"
  >
    <span
      className="title"
    >
      Magic Pages
    </span>
  </div>
  <div
    className="schools-list"
  >
    <MagicSchool
      inEdition={true}
      isActive={true}
      key="0"
      onEditionEnd={[Function]}
      onSelect={[Function]}
      school={
        Object {
          "id": "1",
          "name": "1",
        }
      }
    />
  </div>
  <div
    className="actions-panel"
  >
    <div
      className="action"
      onClick={[Function]}
    >
      Add
    </div>
    <div
      className="action"
      onClick={[Function]}
    >
      Edit
    </div>
    <div
      className="action"
      onClick={[Function]}
    >
      Delete
    </div>
  </div>
</div>
`;

exports[`MagicSchools Renders 1`] = `
<div
  className="schools-block"
>
  <div
    className="block-title"
  >
    <span
      className="title"
    >
      Magic Pages
    </span>
  </div>
  <div
    className="schools-list"
  >
    <MagicSchool
      isActive={false}
      key="0"
      onEditionEnd={[Function]}
      onSelect={[Function]}
      school={
        Object {
          "id": "1",
          "name": "1",
        }
      }
    />
    <MagicSchool
      isActive={false}
      key="1"
      onEditionEnd={[Function]}
      onSelect={[Function]}
      school={
        Object {
          "id": "2",
          "name": "2",
        }
      }
    />
  </div>
  <div
    className="actions-panel"
  >
    <div
      className="action"
      onClick={[Function]}
    >
      Add
    </div>
    <div
      className="action disabled"
      onClick={[Function]}
    >
      Edit
    </div>
    <div
      className="action disabled"
      onClick={[Function]}
    >
      Delete
    </div>
  </div>
</div>
`;
